pkg/client/logs_test.go

```diff
@@ -0,0 +1,185 @@
+	for i := 0; i <= 5; i++ {
+		_, err := reader.Read(mybuf)
+		// This will never be EOF since we never close the channel.
+		if err != nil && err != io.EOF {
+			errcount++
+		}
+	}
+	if errcount == 0 {
+		t.Fatal("did not receive any errors but there should be one.")
+	}
+}
+
+func TestLogReaderNoError(t *testing.T) {
+	testcases := []struct {
+		name          string
+		input         []string
+		bufsize       int
+		expectedReads []string
+	}{
+		{
+			name:          "tiny buffer, simple input",
+			input:         []string{"Hello world 0"},
+			bufsize:       1,
+			expectedReads: []string{"H", "e", "l", "l", "o", " ", "w", "o", "r", "l", "d", " ", "0", ""},
+		},
+		{
+			name:          "small buffer, simple input",
+			input:         []string{"Hello world 0"},
+			bufsize:       2,
+			expectedReads: []string{"He", "ll", "o ", "wo", "rl", "d ", "0"},
+		},
+		{
+			name:          "big buffer, simple input",
+			input:         []string{"Hello world 0"},
+			bufsize:       1000,
+			expectedReads: []string{"Hello world 0"},
+		},
+		{
+			name:          "exact buffer, simple input",
+			input:         []string{"Hello world 0"},
+			bufsize:       len("Hello world 0"),
+			expectedReads: []string{"Hello world 0"},
+		},
+		{
+			name: "big buffer, small messages",
+			input: []string{
+				"Once you start down the dark path, forever will it dominate your destiny.",
+				"Luminous beings are we, not this crude matter.",
+				"Fear is the path to the dark side. Fear leads to anger. Anger leads to hate. Hate leads to suffering.",
+			},
+			bufsize: 1024,
+			expectedReads: []string{
+				"Once you start down the dark path, forever will it dominate your destiny.",
+				"Luminous beings are we, not this crude matter.",
+				"Fear is the path to the dark side. Fear leads to anger. Anger leads to hate. Hate leads to suffering.",
+			},
+		},
+		{
+			name: "small buffer, big input",
+			input: []string{
+				"this is some log line",
+				"this is another log line",
+				"this is a third log line!!",
+			},
+			bufsize: 10,
+			expectedReads: []string{
+				"this is so",
+				"me log lin",
+				"e",
+				"this is an",
+				"other log ",
+				"line",
+				"this is a ",
+				"third log ",
+				"line!!",
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			bytestream := make(chan []byte)
+			go func(data chan []byte) {
+				defer close(bytestream)
+				for _, input := range tc.input {
+					data <- []byte(input)
+				}
+			}(bytestream)
+			errc := make(chan error)
+			reader := NewReader(bytestream, errc)
+			mybuf := make([]byte, tc.bufsize)
+			i := 0
+			for ; ; i++ {
+				n, err := reader.Read(mybuf)
+				if err != nil && err != io.EOF {
+					t.Fatalf("Expected no errors got %v", err)
+				}
+				if err == io.EOF {
+					break
+				}
+				if n > len(mybuf) {
+					t.Fatalf("n is too big: %v mybuf is only %v", n, len(mybuf))
```

Comment:
some of these could probably be `Errorf` instead of `Fatalf`. I usually only use the latter when continuing might lead to a panic (like with the eof)

Reply:
good call, I'll switch them where appropriate